pkg/status/reporter.go

```diff
@@ -0,0 +1,209 @@
+	populators map[BirdConnType]map[apiv3.NodeStatusClassType]statusPopulator
+
+	// channel to indicate this reporter is not needed anymore.
+	// It should start termination process.
+	done chan struct{}
+
+	// channel to indicate this reporter is terminated.
+	term chan struct{}
+}
+
+// newReporter creates a reporter and start running a goroutine handling resource update.
+func newReporter(name string, client client.Interface, intervalInSeconds int, populators map[BirdConnType]map[apiv3.NodeStatusClassType]statusPopulator) *reporter {
+	r := &reporter{
+		name:       name,
+		client:     client,
+		ch:         make(chan *apiv3.CalicoNodeStatus, 10),
+		populators: populators,
+		done:       make(chan struct{}),
+	}
+
+	r.checkAndUpdateTicker(intervalInSeconds)
+
+	go r.run()
+	return r
+}
+
+// Check and set new ticker for the reporter.
+// Make sure stop the old one first to GC old ticker.
+func (r reporter) checkAndUpdateTicker(interval int) {
+	if r.interval == interval {
+		// no update needed.
+		return
+	}
+	if r.ticker != nil {
+		r.ticker.Stop()
+	}
+	r.interval = interval
+	r.ticker = time.NewTicker(time.Duration(interval) * time.Second)
+}
+
+// Cleanup resources owned by this reporter.
+func (r reporter) cleanup() {
+	r.ticker.Stop()
+}
+
+// KillAndWait sends done signal to reporter goroutine and wait until the
+// goroutine of the reporter terminated.
+func (r reporter) KillAndWait() {
+	r.done <- struct{}{}
+	<-r.term
+}
+
+// Called when the caller needs to send a new version of request.
+func (r reporter) RequestUpdate(request *apiv3.CalicoNodeStatus) {
+	r.ch <- request
+}
+
+// run is the main reporting loop, it loops until done.
+func (r reporter) run() {
+	runImmediately := make(chan struct{})
+
+	for {
+		select {
+		case latest := <-r.ch:
+			// Received an update of node status resource.
+			if latest.Name != r.name {
+				log.Warningf("node status reporter (%s) receive request with different name (%s), ignore it", r.name, latest.Name)
+			} else {
+				r.status = latest
+				r.checkAndUpdateTicker(latest.Spec.UpdateIntervalInSeconds)
+				// kick start node status update
+				runImmediately <- struct{}{}
+			}
+		case <-runImmediately:
+		case <-r.ticker.C:
+			if r.status == nil {
+				log.Warningf("No request found for node status reporter (%s)", r.name)
+			} else {
+				r.reportStatus()
+			}
+		case <-r.done:
+			r.cleanup()
+			r.term <- struct{}{}
+			return
+		}
+	}
+}
+
+// reportStatus queries Bird or other components and update node status resource.
+func (r reporter) reportStatus() {
+	var err error
+	// The idea here is that we either update everything successfully or we update nothing.
+
+	// Make a local copy first.
+	status := *r.status
+
+	needUpdate := false
+	for _, ipv := range []BirdConnType{BirdConnTypeV4, BirdConnTypeV6} {
+		failed := false
+		// Populate status from registered populators.
+		for _, class := range r.status.Spec.Classes {
+			if p, ok := r.populators[ipv][class]; ok {
```

Comment:
I'd recommend simplifying the branch structure here as well - make the first branch the `!ok` version and then the rest of the logic doesn't need to be in a branch.

I'm not 100% sure, but I think I'd also be in favor of failing the entire operation if one of the populators fails. I'm not sure what scenarios you were imagining, but I think it would be confusing for users to see an inconsistent set of data (e.g., updated, routes, but not updated peers).

```go
p, ok := r.populators[ipv][class]; 
if !ok {
  log.Warningf("Wrong class (%s) requested for node status reporter (%s)", class, r.name)
  continue
}

err = p.Populate(&status)
if err != nil {
  return err
}
```

Reply:
The only concern I have is that if a cloud platform does not support ipv6 and we are constantly getting errors of that. We will never get update for ipv4. But maybe that is not the case, bird6 works fine even ipv6 is not supported by underlaying infra. We need do some test on it to be sure.

Reply:
Let me simplify the code that we fail on any error. If we see issues regarding bird6, we can come back and fix it.

Reply:
Normally I'm keen on the code structure you outlined above where only error handling is on a branch. Not sure why I didn't do it this time. 😄